README.md
# <img align="left" src="/static/img/icon/by/48.png" alt="T" title="The Tracman Logo">Tracman
###### v 0.9.1

node.js application to display a sharable map with user's location.

[![Travis Build Status](https://travis-ci.org/Tracman-org/Server.svg?branch=develop)](https://travis-ci.org/Tracman-org/Server)
[![Coverage Status](https://coveralls.io/repos/github/Tracman-org/Server/badge.svg?branch=master)](https://coveralls.io/github/Tracman-org/Server?branch=master)
[![JavaScript Style Guide](https://img.shields.io/badge/code_style-standard-brightgreen.svg)](https://standardjs.com)

## Installation

On unix-based systems:

```sh
git clone https://github.com/Tracman-org/Server.git tracman-server &&\
cd tracman-server &&\
npm install
```

You will need to set up a configuration file at `config/env/env.js`.  Use `config/env/sample.js` for an example.  You can get API keys at the [google developer's console](https://console.developers.google.com/apis/credentials).  You will need to set up approved hosts and auth callbacks.  There is more information in [their documentation](https://support.google.com/googleapi/answer/6158857?hl=en).

A good method is to simply copy the sample configuration and point `config/env/env.js` to the new version:

```sh
cp config/env/sample.js config/env/local-config.js
echo "module.exports = require('./local-config.js')" > config/env/env.js
```

Then edit `config/env/local-config.js` to match your local environment.


## Usage

Run Tracman with npm:

```sh
npm run build && npm start
```

...or with [nodemon](https://nodemon.io/):

```sh
npm run nodemon
```

Nodemon will automatically rebuild and restart the app when you make changes. 


## Contributing

Tracman will be updated according to [this branching model](http://nvie.com/posts/a-successful-git-branching-model)... more or less.  If you know anything about programming Android, [the Tracman android app](https://github.com/Tracman-org/Android) is more desperate for help.


## Changelog

[view full changelog](CHANGELOG.md)

###### v0.9.x
* Removed Pro
* Removed coinhive
* Fixed CSP rules
* Improved debugging output
* Tried to fix scrollwheel
* Fixed security audited npm packages
* [#121](https://github.com/Tracman-org/Server/issues/121) Fixed various security holes
* [#68](https://github.com/Tracman-org/Server/issues/68) Added tests, mostly for authentication
* [#120](https://github.com/Tracman-org/Server/issues/120) Split config/routes/settings.js into two files
* Removed express validator and replaced with homegrown function
* Fixed showing welcome message on every login
* Removed naked domains
* Patched [CVE-2017-16117](https://github.com/dodo/node-slug/issues/82) on [node-slug](https://github.com/dodo/node-slug)

###### v0.8.x
* Hotfixed service worker bugs
* Added check to ensure only the newest location is sent
* Removed buggy login/-out redirects
* [#111](https://github.com/Tracman-org/Server/issues/111) Implemented service worker
* [#116](https://github.com/Tracman-org/Server/issues/116) Switched promises for async/await
* [#64](https://github.com/Tracman-org/Server/issues/64) Started using promises in model methods
* [#63](https://github.com/Tracman-org/Server/issues/63) Disconnect inactive clients
* [#110](https://github.com/Tracman-org/Server/issues/110) Implemented [StandardJS](https://standardjs.com/)
* Removed emojis
* [#113](https://github.com/Tracman-org/Server/issues/113) [#114](https://github.com/Tracman-org/Server/issues/114) Switched to useMongoClient
* [#67](https://github.com/Tracman-org/Server/issues/67) Fixed invisible marker in demo
* Shortened if statements and long lines

#### v0.7.x
* Fixed altitude sign
* Added coinhive
* Added HTML5 shiv
* Added support for browser geolocation
* Fixed streetview image bugs
* Added sourcemaps for debugging
* Fixed streetview bearing
* [#96](https://github.com/Tracman-org/Server/issues/96) Replaced panorama with static images
* [#96](https://github.com/Tracman-org/Server/issues/96) Continue to update panorama after creating element
* [#100](https://github.com/Tracman-org/Server/issues/100) Moved CSS out of webpack
* Fixed [#108](https://github.com/Tracman-org/Server/issues/108)
* Fixed bugs [#105](https://github.com/Tracman-org/Server/issues/105) and [#107](https://github.com/Tracman-org/Server/issues/107)
* [#89](https://github.com/Tracman-org/Server/issues/89) Can change marker color
* Fixed [#102](https://github.com/Tracman-org/Server/issues/102) issue creating account


## License

[view full license](LICENSE.md)

Tracman: GPS tracking service in node.js
Copyright © 2018 [Keith Irwin](https://www.keithirwin.us/)

This program is free software: you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation, either version 3 of the License, or (at your option) any later version.

This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more details.

You should have received a copy of the GNU General Public License along with this program.  If not, see [http://www.gnu.org/licenses/](http://www.gnu.org/licenses/).


CHANGELOG.md
# Tracman Server Changelog
### v0.9.1

###### v0.9.1
* Removed Pro
* Removed conhive
* Fixed CSP rules
* Improved debugging output
* Tried to fix scrollwheel
* Fixed security audited npm packages
* Patched [CVE-2017-16117](https://github.com/dodo/node-slug/issues/82) on [node-slug](https://github.com/dodo/node-slug)

###### v0.9.0
* [#121](https://github.com/Tracman-org/Server/issues/121) Fixed various security holes
* [#68](https://github.com/Tracman-org/Server/issues/68) Added tests, mostly for authentication
* [#120](https://github.com/Tracman-org/Server/issues/120) Split config/routes/settings.js into two files
* Removed express validator and replaced with homegrown function
* Fixed showing welcome message on every login
* Removed naked domains

###### v0.8.1/2
* Hotfixed service worker bugs

###### v0.8.0
* Added check to ensure only the newest location is sent
* Removed buggy login/-out redirects
* [#111](https://github.com/Tracman-org/Server/issues/111) Implemented service worker
* [#116](https://github.com/Tracman-org/Server/issues/116) Switched promises for async/await
* [#64](https://github.com/Tracman-org/Server/issues/64) Started using promises in model methods
* [#63](https://github.com/Tracman-org/Server/issues/63) Disconnect inactive clients
* [#110](https://github.com/Tracman-org/Server/issues/110) Implemented [StandardJS](https://standardjs.com/)
* Removed emojis
* [#113](https://github.com/Tracman-org/Server/issues/113) [#114](https://github.com/Tracman-org/Server/issues/114) Switched to useMongoClient
* [#67](https://github.com/Tracman-org/Server/issues/67) Fixed invisible marker in demo
* Shortened if statements and long lines

#### v0.7.12
* Fixed altitude sign

#### v0.7.10
* Added coinhive
* Updated jQuery
* Added HTML5 shiv

#### v0.7.9
* Updated packages
* Added support for browser geolocation
* Fixed streetview image bugs
* Added sourcemaps for debugging
* Fixed streetview bearing

#### v0.7.8
* [#96](https://github.com/Tracman-org/Server/issues/96) Replaced panorama with static images

#### v0.7.7
* [#96](https://github.com/Tracman-org/Server/issues/96) Continue to update panorama after creating element

#### v0.7.6
* [#100](https://github.com/Tracman-org/Server/issues/100) Moved CSS out of webpack
* Fixed [#108](https://github.com/Tracman-org/Server/issues/108)

#### v0.7.5
* Added pricing to homepage

#### v0.7.4
* Improved debugging
* Fixed bugs [#105](https://github.com/Tracman-org/Server/issues/105) and [#107](https://github.com/Tracman-org/Server/issues/107)
* [#89](https://github.com/Tracman-org/Server/issues/89) Can change marker color

#### v0.7.3
* Fixed [#102](https://github.com/Tracman-org/Server/issues/102) issue creating account

#### v0.7.2
* Added more debugging to fix auth problems

#### v0.7.0
* More bug fixes

#### v0.6.7
* [#87](https://github.com/Tracman-org/Server/issues/87) Started loading assets with webpack

#### v0.6.6

* Removed demo recording code
* Moved email server settings to env file
* Added SMTP check

#### v0.6.5

* [#96](https://github.com/Tracman-org/Server/issues/96) Fixed Google streetview
* Various minor fixes

#### v0.6.4

* [#92](https://github.com/Tracman-org/Server/issues/92) Fixed blank map issue

#### v0.6.3

* [#84](https://github.com/Tracman-org/Server/issues/84) Fixed password reset page

#### v0.6.2

* [#76](https://github.com/Tracman-org/Server/issues/76) Prevented users with no password from deleting google social login
* Fixed error when creating user

#### v0.6.1

* [#77](https://github.com/Tracman-org/Server/issues/77) Fixed 500 after password change, swapped `bcrypt` for `bcrypt-nodejs`
* Removed extraneous packages

#### v0.6.0

* [#32](https://github.com/Tracman-org/Server/issues/32), [#57](https://github.com/Tracman-org/Server/issues/57), [#58](https://github.com/Tracman-org/Server/issues/58), [#60](https://github.com/Tracman-org/Server/issues/60) Added more login options
* [#50](https://github.com/Tracman-org/Server/issues/50) Replaced some callbacks with promises
* Minified static files
* [#51](https://github.com/Tracman-org/Server/issues/51), [#52](https://github.com/Tracman-org/Server/issues/52) Added settings validations
* [#54](https://github.com/Tracman-org/Server/issues/54), [#55](https://github.com/Tracman-org/Server/issues/55) Made map work better
* [#61](https://github.com/Tracman-org/Server/issues/61) New MongoDB security
* [#62](https://github.com/Tracman-org/Server/issues/62) Fixed error handling

#### v0.5.1

* Fixed broken controls

#### v0.5.0

* Updated libraries
* Fixed recognition of attached clients [#34](https://github.com/Tracman-org/Server/issues/21)
* Moved socket.io code to own file.
* Many minor fixes

#### v0.4.3

* Fixed memory store [#21](https://github.com/Tracman-org/Server/issues/21)

#### v0.4.2

* Fixed Streetview covering buttons
* Fixed error when viewing map of nonexistant user

#### v0.4.1

* Users can view/change email address
* Added linked accounts to admin

#### v0.4.0

* Opened registration
* Replaced 'Imperial' with 'Standard'
* Bug fixes

#### v0.3.0

* Unified map and dashboard UI
* Security updates
* New admin UI


server.js
'use strict'

/* IMPORTS */
const express = require('express')
const helmet = require('helmet')
const csp = require('helmet-csp')
const rateLimit = require('express-request-limit')
const bodyParser = require('body-parser')
const cookieParser = require('cookie-parser')
const cookieSession = require('cookie-session')
const csurf = require('csurf')
const mongoose = require('mongoose')
const nunjucks = require('nunjucks')
const passport = require('passport')
const flash = require('connect-flash-plus')
const env = require('./config/env/env')
const User = require('./config/models').user
const Map = require('./config/models').map
const mail = require('./config/mail')
const demo = require('./config/demo')
const app = express()
const rescheme = require('./config/rescheme')
const http = require('http').Server(app)
const io = require('socket.io')(http)
const sockets = require('./config/sockets')
const square = require('square-connect')
const debug = require('debug')('tracman-server')


// Set up square
square.ApiClient.instance.authentications['oauth2'].accessToken = env.squareAccessToken
const squareApi = new square.LocationsApi()

// Promises marking a ready server
const ready_promise_list = []

// Show stacktrace for unhandled promise rejections
// https://github.com/nodejs/node/issues/9523#issuecomment-259303079
if (env.mode !== 'production')
  process.on('unhandledRejection', (err) => console.error(err));

/* Database */ {
  // Setup with native ES6 promises
  mongoose.Promise = global.Promise

  // Connect to database
  ready_promise_list.push( new Promise( async (resolve, reject) => {
    try {
      mongoose.connect(env.mongoSetup, {
        useMongoClient: true,
        socketTimeoutMS: 30000,
        //reconnectTries: 30,
        keepAlive: true
      })
      console.log(`  Mongoose connected to ${env.mongoSetup}`)
      resolve()
    } catch (err) {
      console.error(err.stack)
      reject()
    }
  }) )

}

/* Connect square */ {
  
  ready_promise_list.push( new Promise( async (resolve, reject) => {
    try {
      console.log(`  Square API called successfully. `)
      resolve()
    } catch (err) {
      console.error(err.stack)
      reject()
    }
  }) )
  
}

/* Templates */ {
  nunjucks.configure(__dirname + '/views', {
    autoescape: true,
    express: app,
  })
  app.set('view engine', 'html')
}

/* Express session and settings */  app.use(
  helmet.referrerPolicy({
    policy: 'strict-origin',
  }),
  csp({directives:{
    'default-src': ["'self'"],
    'script-src': ["'self'",
      "'unsafe-inline'", // TODO: Get rid of this
      'https://code.jquery.com',
      'https://cdnjs.cloudflare.com/ajax/libs/socket.io/',
      'https://cdnjs.cloudflare.com/ajax/libs/moment.js/',
      'https://www.google.com/recaptcha/',
      'https://www.google-analytics.com',
      'https://maps.googleapis.com',
      // 'https://coin-hive.com',
      // 'https://coinhive.com',
    ],
    // 'worker-src': ["'self'",
    //   'blob:', // for coinhive
    // ],
    'connect-src': ["'self'",
      'wss://*.tracman.org',
      'wss://tracman-keith24.c9users.io',
      'https://www.google-analytics.com',
      // 'wss://*.coinhive.com',
    ],
    'style-src': ["'self'",
      "'unsafe-inline'",
      'https://fonts.googleapis.com',
      'https://maxcdn.bootstrapcdn.com',
    ],
    'font-src': [
      'https://fonts.gstatic.com',
      'https://maxcdn.bootstrapcdn.com/font-awesome/*',
      'https://maxcdn.bootstrapcdn.com/font-awesome/4.5.0/fonts/fontawesome-webfont.woff',
      'https://maxcdn.bootstrapcdn.com/font-awesome/4.5.0/fonts/fontawesome-webfont.ttf',
      'https://maxcdn.bootstrapcdn.com/font-awesome/4.5.0/fonts/fontawesome-webfont.woff2',
    ],
    'img-src': ["'self'",
      'data:', // For icons on map
      'https://www.google-analytics.com',
      'https://maps.gstatic.com',
      'https://*.googleapis.com',
      'https://http.cat',
    ],
    'object-src': ["'none'"],
    'report-uri': '/csp-violation',
  }}),
  cookieParser(env.cookie),
  cookieSession({
    cookie: {
      maxAge: 1000 * 60 * 60 * 24 * 7, // 1 week
      secure: true,
      httpOnly: true,
      domain: env.url.substring(env.url.indexOf('//')+2),
    },
    secret: env.session,
    saveUninitialized: true,
    resave: true,
  }),
  bodyParser.json(),
  bodyParser.urlencoded({
    extended: true,
  }),
  flash()
)

/* Report CSP violations */
app.post('/csp-violation', (req, res) => {
  console.log(`CSP Violation: ${JSON.stringify(req.body)}`)
  res.status(204).end()
})

/* Auth */ {
  require('./config/passport')(passport)
  app.use(passport.initialize(), passport.session())
}

/* Routes  */ {
  // Static files (keep this before default locals)
  app.use('/static', express.static(__dirname + '/static', {dotfiles: 'allow'}))

  // Default locals available to all views (keep this after static files)
  app.get('*', (req, res, next) => {

    // Rate limit
    rateLimit({
      timeout: 1000 * 60 * 30, // 30 minutes
      exactPath: true,
      cleanUpInterval: 1000 * 60 * 60 * 24 * 7, // 1 week
    })

    // User account
    res.locals.user = req.user

    // Flash messages
    res.locals.successes = req.flash('success')
    res.locals.dangers = req.flash('danger')
    res.locals.warnings = req.flash('warning')

    next()
  })

  // Auth routes
  require('./config/routes/auth')(app, passport)

  // Main routes
  app.use('/', require('./config/routes/index'))

  // Contact form
  app.use('/contact', require('./config/routes/contact'))

  // Settings
  app.use('/settings', require('./config/routes/settings'))

  // Maps
  app.use('/maps', require('./config/routes/maps'))

  // Account settings
  app.use('/account', require('./config/routes/account'))

  // Map
  app.use(['/map', '/trac'], require('./config/routes/map'))

  // Site administration
  app.use('/admin', require('./config/routes/admin'))

  // Testing
  if (env.mode == 'development') app.use('/test', require('./config/routes/test'))

} {

  // Catch-all for 404s
  app.use((req, res, next) => {
    if (!res.headersSent) {
      const err = Error(`Not found: ${req.url}`)
      err.status = 404
      next(err)
    }
  })

  // Production handlers
  if (env.mode !== 'development') {
    app.use((err, req, res, next) => {
      if (err.status >= 500) console.error(err.stack)
      if (res.headersSent) return next(err)
      res.status(err.status || 500)
      res.render('error', {
        code: err.status || 500,
        message: (err.status < 500) ? err.message : 'Server error'
      })
    })

  // Development handlers
  } else {
    app.use((err, req, res, next) => {
      if (err.status !== 404) console.error(err.stack)
      if (res.headersSent) return next(err)
      res.status(err.status || 500)
      res.render('error', {
        code: err.status || 500,
        message: err.message,
        stack: err.stack
      })
    })
  }
}

// CSRF Protection (keep after routes)
app.use(csurf({
  cookie: true,
}))

/* Sockets */
sockets.init(io)

/* RUNTIME */
console.log(`Starting ${env.mode} server at ${__dirname}...`)

// Test SMTP server
ready_promise_list.push(mail.verify())

// Rescheme database
//TODO: Remove this after reschemed
ready_promise_list.push( new Promise( async (resolve, reject) => {
  try {
    let old_users = await User.find({})
    let old_user_ids = []
    old_users.forEach( (old_user) => {
      old_user_ids.push(old_user._id)
    })
    old_user_ids.forEach((old_user_id) => {
      rescheme(old_user_id)
    })
    resolve()
  } catch (err) {
    console.error(err.message)
    reject(err)
  }
}) )

// Delete users who never created a password
// TODO: Uncomment after rescheme
// ready_promise_list.push( new Promise( async (resolve, reject) => {
//   try {

//     // Get users with no passwords
//     (await User.find({
//       'auth.password': {$exists:false},
//     })

//     // Check that their token expired too
//     .where('auth.passTokenExpires').lt(Date.now())

//     // Delete them one-by-one to ensure hooks fire
//     ).forEach( (user) => {
//       debug(`Deleting abandoned account ${user.id}...`)
//       user.remove()
//     })

//     resolve()
//   } catch (err) {
//     console.error(err.message)
//     reject(err)
//   }
// }) )

// Check for spectators for all users
ready_promise_list.push( new Promise( async (resolve, reject) => {
  try {
    (await Map.find({})).forEach( (map) => {
      sockets.checkForUsers(io, map.id)
    })
    resolve()
  } catch (err) {
    console.error(err.message)
    reject(err)
  }
}) )

// Start transmitting demo
ready_promise_list.push( demo(io) )

// Listen
ready_promise_list.push( new Promise( (resolve, reject) => {
  http.listen(env.port, () => {
    console.log(`  Express listening on ${env.url}`)
    resolve()
  })
}) )

// Mark everything when working correctly
Promise.all(ready_promise_list.map(
  // Also wait for rejected promises
  // https://stackoverflow.com/a/36115549/3006854
  p => p.catch(e => e)
)).then( () => {
  console.log('Tracman server is running properly.\n')
}).catch( (err) => {
  console.error(err.message)
  console.log(`Tracman server is NOT running properly!\n`)
})

module.exports = app
